libc/string/string_lenchr.c

```diff
@@ -52,105 +54,303 @@ TEST(string_len, ascii)
+		TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 1, strcspn(&asciiStr[1], &reversCharSet[i]));
+
+		if (i == 127) {
+			TEST_ASSERT_EQUAL_INT(BUFF_SIZE - 1, strspn(&asciiStr[1], reversCharSet));
+		}
+		else {
+			TEST_ASSERT_EQUAL_INT(0, strspn(&asciiStr[1], reversCharSet));
+		}
+	}
+
+	free((void *)asciiStr);
+}
+
+
+TEST(string_spn, not_ascii)
+{
+	unsigned char charSet[BUFF_SIZE] = { 0 };
+	int len, i;
+
+	/* Checking  out of ASCII bytes */
+	for (i = sizeof(charSet); i < sizeof(charSet) * 2 - 1; i++) {
+		charSet[i - sizeof(charSet)] = i;
 	}
-	bigstr[i] = '\0';
 
-	sz = sizeof(bigstr) - 1;
-	TEST_ASSERT_EQUAL_INT(sz, strlen(bigstr));
-	TEST_ASSERT_EQUAL_INT(sz - 1, strnlen(bigstr, sz - 1));
-	TEST_ASSERT_EQUAL_INT(sz, strnlen(bigstr, sz));
-	TEST_ASSERT_EQUAL_INT(sz, strnlen(bigstr, sz + 1));
+	len = sizeof(charSet) - 1;
+	TEST_ASSERT_EQUAL_INT(len, strcspn((const char *)charSet, ""));
+	TEST_ASSERT_EQUAL_INT(len, strspn((const char *)charSet, (const char *)charSet));
+}
+
+
+TEST(string_spn, big)
+{
+	char bigstr[PATH_MAX] = { 0 },
+		 supportStr[PATH_MAX] = { 0 };
+	int len;
+
+	/*
+	 * The length of the string is not restricted,
+	 * so we test one of the bigger values, which may be used
+	 * with remembering the last element must be a null term
+	 * zero to collect the data as the string from
+	 * declared space, that's why we memset one place less than its size
+	 */
+
+	len = sizeof(bigstr) - 1;
+	memset(bigstr, 'a', len);
+
+	TEST_ASSERT_EQUAL_INT(len, strcspn(bigstr, ""));
+	TEST_ASSERT_EQUAL_INT(0, strcspn(bigstr, "a"));
+	TEST_ASSERT_EQUAL_INT(len, strspn(bigstr, "a"));
+	TEST_ASSERT_EQUAL_INT(0, strspn(bigstr, ""));
+
+	memcpy(supportStr, bigstr, len);
+
+	TEST_ASSERT_EQUAL_INT(len, strcspn(supportStr, ""));
+	TEST_ASSERT_EQUAL_INT(0, strspn(supportStr, ""));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn(supportStr, "a"));
+	TEST_ASSERT_EQUAL_INT(len, strspn(supportStr, "a"));
+
+	TEST_ASSERT_EQUAL_INT(len, strcspn(supportStr, "b"));
+	TEST_ASSERT_EQUAL_INT(0, strspn(supportStr, "b"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn(supportStr, "ab"));
+	TEST_ASSERT_EQUAL_INT(len, strspn(supportStr, "ab"));
+
+	supportStr[len - 3] = 'b';
+	TEST_ASSERT_EQUAL_INT(len - 3, strspn(supportStr, bigstr));
+	supportStr[len - 3] = 'a';
+	TEST_ASSERT_EQUAL_INT(len, strspn(supportStr, bigstr));
+
+	memset(supportStr, 'b', len);
+	supportStr[len - 3] = 'a';
+	TEST_ASSERT_EQUAL_INT(len - 3, strcspn(supportStr, bigstr));
+	supportStr[len - 3] = 'b';
+	TEST_ASSERT_EQUAL_INT(len, strcspn(supportStr, bigstr));
+}
+
+
+TEST(string_spn, empty_args)
+{
+	TEST_ASSERT_EQUAL_INT(0, strcspn("", "abc"));
+	TEST_ASSERT_EQUAL_INT(0, strspn("", "abc"));
+
+	TEST_ASSERT_EQUAL_INT(0, strcspn("", ""));
+	TEST_ASSERT_EQUAL_INT(0, strspn("", ""));
+
+	TEST_ASSERT_EQUAL_INT(3, strcspn("abc", ""));
+	TEST_ASSERT_EQUAL_INT(0, strspn("abc", ""));
+
+	TEST_ASSERT_EQUAL_INT(1, strcspn("abc", "bc"));
+	TEST_ASSERT_EQUAL_INT(0, strspn("abc", "bc"));
+
```

Comment:
As you can see, there are no empty arguments, but IMHO it's a good case to check. Let's create another test case and name it `first_byte`, or sth like that, then I'd suggest: `strcspn("abc", "a")` instead of `strcspn("abc", "bc")` to check scenarios where initial byte shouldn't be counted.

It would be good to add a comment to this test case, sth like: `This case is intended to check the ability of stopping at first byte`

You can add more checks, like:
 - `strspn("abc", "c")`
 - `strcspn("abc", "ab")`
 - `strspn("aaaabc", "bc")`
 - `strcspn("aaaabc", "a")`

Reply:
Most of the mentioned cases haven't been applied. All of checks there should return 0 as we check this utility (stopping at once).